RFE: improve message handling when getting HTTP/401 from AS3

#### Title
RFE: improve message handling when getting HTTP/401 from AS3 

#### Description

During a customer PoC, I´ve found that there was a an authentication error in CIS, only the following was being sent:

```
$ oc logs k8s-bigip1
-ctlr-deployment-79d695b5ff-nsvjk
2024/03/22 12:30:16 [INFO] [INIT] Starting: Container Ingress Services - Version: 2.16.0, BuildInfo: azure-5835-34b43c48e8d0f41b0d6d0b80644027a9c8286dc3
2024/03/22 12:30:16 [INFO] ConfigWriter started: 0xc00074c750
2024/03/22 12:30:16 [WARNING] Creating GTM with default bigip credentials as GTM BIGIP Url or GTM BIGIP Username or GTM BIGIP Password is missing on CIS args.
2024/03/22 12:30:16 [INFO] Started config driver sub-process at pid: 18
2024/03/22 12:30:16 [INFO] [CORE] Registered BigIP Metrics
2024/03/22 12:30:16 [ERROR] [AS3] Error response from BIGIP with status code 401
2024/03/22 12:30:16 [ERROR] Error response from BIGIP with status code 401
2024/03/22 12:30:16 [INFO] [CCCL] ConfigWriter stopped: 0xc00074c750
```
where it is not clear at all what is going on, while if sending a request with POSTMAN the following is retrieved:

```
{
    "code": 401,
    "message": "Authentication failed: Password expired. Update password via /mgmt/shared/authz/users.",
    "restOperationId": 4501183,
    "kind": ":resterrorresponse"
}
```

where it is very clear what the problem is. I wonder if we could incorporate this in our logs.

Many thanks

        Ulises

--- a/docs/RELEASE-NOTES.rst
+++ b/docs/RELEASE-NOTES.rst
@@ -22,6 +22,7 @@ Bug Fixes
 * `Issue 3466 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3466>`_: Fix scaling issue in NodeportLocal mode
 * `Issue 3432 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3432>`_: Show meaningful logs for exceptions occured from controller agent
 * `Issue 3396 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3396>`_: Fix adding pool members from external clusters in nodeportLocal mc mode
+* `Issue 3351 <https://github.com/F5Networks/k8s-bigip-ctlr/issues/3351>`_: improve message handling when getting HTTP/401 from AS3
 
 Upgrade notes
 ``````````````

--- a/pkg/agent/as3/postManager.go
+++ b/pkg/agent/as3/postManager.go
@@ -46,6 +46,7 @@ const (
 	responseStatusNotFound            = "statusNotFound"
 	responseStatusServiceUnavailable  = "statusServiceUnavailable"
 	responseStatusUnprocessableEntity = "statusUnprocessableEntity"
+	responseStatusUnAuthorized        = "statusUnAuthorized"
 	responseStatusDummy               = "dummy"
 )
 
@@ -183,6 +184,8 @@ func (postMgr *PostManager) postConfigRequests(data string, url string) (bool, s
 		return postMgr.handleResponseStatusNotFound(responseMap)
 	case http.StatusUnprocessableEntity:
 		return postMgr.handleStatusUnprocessableEntity(responseMap)
+	case http.StatusUnauthorized:
+		return postMgr.handleResponseStatusUnAuthorized(responseMap)
 	default:
 		return postMgr.handleResponseOthers(responseMap)
 	}
@@ -220,6 +223,18 @@ func (postMgr *PostManager) GetBigipAS3Version() (string, string, string, error)
 		}
 		// In case of 503 status code : CIS will exit and auto restart of the
 		// controller might fetch the BIGIP version once BIGIP is available.
+	case http.StatusUnauthorized:
+		if code, ok := responseMap["code"].(float64); ok {
+			if int(code) == http.StatusUnauthorized {
+				if _, ok := responseMap["message"].(string); ok {
+					return "", "", "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v Message: %v", httpResp.StatusCode, responseMap["message"])
+				} else {
+					return "", "", "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v", httpResp.StatusCode)
+				}
+			}
+		}
 	}
 	return "", "", "", fmt.Errorf("Error response from BIGIP with status code %v", httpResp.StatusCode)
 }
@@ -252,6 +267,18 @@ func (postMgr *PostManager) GetBigipRegKey() (string, error) {
 			return "", fmt.Errorf("AS3 RPM is not installed on BIGIP,"+
 				" Error response from BIGIP with status code %v", httpResp.StatusCode)
 		}
+	case http.StatusUnauthorized:
+		if code, ok := responseMap["code"].(float64); ok {
+			if int(code) == http.StatusUnauthorized {
+				if _, ok := responseMap["message"].(string); ok {
+					return "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v Message: %v", httpResp.StatusCode, responseMap["message"])
+				} else {
+					return "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v", httpResp.StatusCode)
+				}
+			}
+		}
 	}
 	return "", fmt.Errorf("Error response from BIGIP with status code %v", httpResp.StatusCode)
 }
@@ -273,6 +300,9 @@ func (postMgr *PostManager) httpReq(request *http.Request) (*http.Response, map[
 	err = json.Unmarshal(body, &response)
 	if err != nil {
 		log.Errorf("[AS3] Response body unmarshal failed: %v\n", err)
+		if httpResp.StatusCode == http.StatusUnauthorized {
+			log.Errorf("[AS3] Unauthorized access to BIG-IP, please check the credentials, message: %v", string(body))
+		}
 		if postMgr.LogAS3Response {
 			log.Errorf("[AS3] Raw response from Big-IP: %v", string(body))
 		}
@@ -313,6 +343,23 @@ func (postMgr *PostManager) handleResponseStatusNotFound(responseMap map[string]
 		postMgr.logAS3Response(responseMap)
 	}
 	return true, responseStatusNotFound
+}
+
+func (postMgr *PostManager) handleResponseStatusUnAuthorized(responseMap map[string]interface{}) (bool, string) {
+	if _, ok := responseMap["code"].(float64); ok {
+		if _, ok := responseMap["message"].(string); ok {
+			log.Errorf("[AS3] authentication failed,"+
+				" Error response from BIGIP with status code: 401 Message: %v", responseMap["message"])
+		} else {
+			log.Errorf("[AS3] authentication failed," +
+				" Error response from BIGIP with status code: 401")
+		}
+	}
+
+	if postMgr.LogAS3Response {
+		postMgr.logAS3Response(responseMap)
+	}
+	return true, responseStatusUnAuthorized
 }
 
 func (postMgr *PostManager) handleStatusUnprocessableEntity(responseMap map[string]interface{}) (bool, string) {

--- a/pkg/controller/postManager.go
+++ b/pkg/controller/postManager.go
@@ -158,6 +158,9 @@ func (postMgr *PostManager) postConfig(cfg *agentConfig) {
 	case http.StatusNotFound:
 		log.Infof("%v[AS3]%v post resulted in FAILURE", getRequestPrefix(cfg.id), postMgr.postManagerPrefix)
 		postMgr.handleResponseStatusNotFound(responseMap, cfg.id)
+	case http.StatusUnauthorized:
+		log.Infof("%v[AS3]%v post resulted in UNAUTHORIZED FAILURE", getRequestPrefix(cfg.id), postMgr.postManagerPrefix)
+		postMgr.handleResponseStatusUnAuthorized(responseMap, cfg.id)
 	default:
 		log.Infof("%v[AS3]%v post resulted in FAILURE", getRequestPrefix(cfg.id), postMgr.postManagerPrefix)
 		postMgr.handleResponseOthers(responseMap, cfg.id)
@@ -190,6 +193,9 @@ func (postMgr *PostManager) httpPOST(request *http.Request) (*http.Response, map
 	err = json.Unmarshal(body, &response)
 	if err != nil {
 		log.Errorf("[AS3]%v Response body unmarshal failed: %v\n", postMgr.postManagerPrefix, err)
+		if httpResp.StatusCode == http.StatusUnauthorized {
+			log.Errorf("[AS3]%v Unauthorized access to BIG-IP, please check the credentials, message: %v", postMgr.postManagerPrefix, string(body))
+		}
 		if postMgr.LogAS3Response {
 			log.Errorf("[AS3]%v Raw response from Big-IP: %v", postMgr.postManagerPrefix, string(body))
 		}
@@ -363,6 +369,26 @@ func (postMgr *PostManager) handleResponseStatusNotFound(responseMap map[string]
 	postMgr.updateTenantResponseCode(http.StatusNotFound, "", "", false)
 }
 
+func (postMgr *PostManager) handleResponseStatusUnAuthorized(responseMap map[string]interface{}, id int) {
+	var unknownResponse bool
+	if _, ok := responseMap["code"].(float64); ok {
+		if _, ok := responseMap["message"].(string); ok {
+			log.Errorf("%v[AS3]%v authentication failed,"+
+				" Error response from BIGIP with status code: 401 Message: %v", getRequestPrefix(id), postMgr.postManagerPrefix, responseMap["message"])
+		} else {
+			log.Errorf("%v[AS3]%v authentication failed,"+
+				" Error response from BIGIP with status code: 401", getRequestPrefix(id), postMgr.postManagerPrefix)
+		}
+	} else {
+		unknownResponse = true
+	}
+
+	if postMgr.LogAS3Response || unknownResponse {
+		postMgr.logAS3Response(responseMap)
+	}
+	postMgr.updateTenantResponseCode(http.StatusUnauthorized, "", "", false)
+}
+
 func (postMgr *PostManager) handleResponseOthers(responseMap map[string]interface{}, id int) {
 	var unknownResponse bool
 	if results, ok := (responseMap["results"]).([]interface{}); ok {
@@ -440,6 +466,20 @@ func (postMgr *PostManager) GetBigipAS3Version() (string, string, string, error)
 		}
 		// In case of 503 status code : CIS will exit and auto restart of the
 		// controller might fetch the BIGIP version once BIGIP is available.
+	case http.StatusUnauthorized:
+		if code, ok := responseMap["code"].(float64); ok {
+			if int(code) == http.StatusUnauthorized {
+				if _, ok := responseMap["message"].(string); ok {
+					return "", "", "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v Message: %v", httpResp.StatusCode, responseMap["message"])
+				} else {
+					return "", "", "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v", httpResp.StatusCode)
+				}
+			}
+		} else {
+			unknownResponse = true
+		}
 	}
 	if unknownResponse {
 		postMgr.logAS3Response(responseMap)
@@ -484,6 +524,20 @@ func (postMgr *PostManager) GetBigipRegKey() (string, error) {
 		} else {
 			unknownResponse = true
 		}
+	case http.StatusUnauthorized:
+		if code, ok := responseMap["code"].(float64); ok {
+			if int(code) == http.StatusUnauthorized {
+				if _, ok := responseMap["message"].(string); ok {
+					return "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v Message: %v", httpResp.StatusCode, responseMap["message"])
+				} else {
+					return "", fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v", httpResp.StatusCode)
+				}
+			}
+		} else {
+			unknownResponse = true
+		}
 	}
 	if unknownResponse {
 		postMgr.logAS3Response(responseMap)
@@ -519,6 +573,20 @@ func (postMgr *PostManager) GetAS3DeclarationFromBigIP() (map[string]interface{}
 		} else {
 			postMgr.logAS3Response(responseMap)
 		}
+	case http.StatusUnauthorized:
+		if code, ok := responseMap["code"].(float64); ok {
+			if int(code) == http.StatusUnauthorized {
+				if _, ok := responseMap["message"].(string); ok {
+					return nil, fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v Message: %v", httpResp.StatusCode, responseMap["message"])
+				} else {
+					return nil, fmt.Errorf("authentication failed,"+
+						" Error response from BIGIP with status code %v", httpResp.StatusCode)
+				}
+			}
+		} else {
+			postMgr.logAS3Response(responseMap)
+		}
 	}
 	return nil, fmt.Errorf("Error response from BIGIP with status code %v", httpResp.StatusCode)
 }
@@ -540,6 +608,9 @@ func (postMgr *PostManager) httpReq(request *http.Request) (*http.Response, map[
 	err = json.Unmarshal(body, &response)
 	if err != nil {
 		log.Errorf("[AS3]%v Response body unmarshal failed: %v\n", postMgr.postManagerPrefix, err)
+		if httpResp.StatusCode == http.StatusUnauthorized {
+			log.Errorf("[AS3]%v Unauthorized access to BIG-IP, please check the credentials, message: %v", postMgr.postManagerPrefix, string(body))
+		}
 		if postMgr.LogAS3Response {
 			log.Errorf("[AS3]%v Raw response from Big-IP: %v", postMgr.postManagerPrefix, string(body))
 		}

